crypto/compactcert/builder.go

```diff
@@ -0,0 +1,244 @@
+	b.sigs[pos].part = p
+	b.sigs[pos].Sig.OneTimeSignature = sig
+	b.signedWeight += p.Weight
+	b.cert = nil
+	return nil
+}
+
+// Ready returns whether the certificate is ready to be built.
+func (b *Builder) Ready() bool {
+	return b.signedWeight >= b.Params.ProvenWeight
+}
+
+// SignedWeight returns the total weight of signatures added so far.
+func (b *Builder) SignedWeight() uint64 {
+	return b.signedWeight
+}
+
+//msgp:ignore sigCommit
+type sigCommit []sigslot
+
+func (sc sigCommit) Length() uint64 {
+	return uint64(len(sc))
+}
+
+func (sc sigCommit) Get(pos uint64) (crypto.Hashable, error) {
+	if pos >= uint64(len(sc)) {
+		return nil, fmt.Errorf("pos %d past end %d", pos, len(sc))
+	}
+
+	return &sc[pos].sigslotCommit, nil
+}
+
+// coinIndex returns the position pos in the sigs array such that the sum
+// of all signature weights before pos is less than or equal to coinWeight,
+// but the sum of all signature weights up to and including pos exceeds
+// coinWeight.
+//
+// coinIndex works by doing a binary search on the sigs array.  The caller
+// should invoke it with lo=0 and hi=len(b.sigs).  The caller should make
+// sure that sigs[*].L is initialized before using coinIndex().
+func (b *Builder) coinIndex(coinWeight uint64, lo uint64, hi uint64) (uint64, error) {
+again:
+	if lo >= hi {
+		return 0, fmt.Errorf("coinIndex: lo %d >= hi %d", lo, hi)
+	}
+
+	mid := (lo + hi) / 2
+	if coinWeight < b.sigs[mid].L {
+		hi = mid
+		goto again
+	}
+
+	if coinWeight < b.sigs[mid].L+b.sigs[mid].part.Weight {
+		return mid, nil
+	}
+
+	lo = mid + 1
+	goto again
+}
+
+// Build returns a compact certificate, if the builder has accumulated
+// enough signatures to construct it.
+func (b *Builder) Build() (*Cert, error) {
+	if b.cert != nil {
+		return b.cert, nil
+	}
+
+	if b.signedWeight < b.Params.ProvenWeight {
+		return nil, fmt.Errorf("not enough signed weight: %d < %d", b.signedWeight, b.Params.ProvenWeight)
+	}
+
+	// Commit to the sigs array
+	for i := 1; i < len(b.sigs); i++ {
+		b.sigs[i].L = b.sigs[i-1].L + b.sigs[i-1].part.Weight
+	}
+
+	sigtree, err := merklearray.Build(sigCommit(b.sigs))
+	if err != nil {
+		return nil, err
+	}
+
+	// Reveal sufficient number of signatures
+	c := &Cert{
+		SigCommit:    sigtree.Root(),
+		SignedWeight: b.signedWeight,
+		Reveals:      nil,
+	}
+
+	nr, err := b.numReveals(b.signedWeight)
+	if err != nil {
+		return nil, err
+	}
+
+	var proofPositions []uint64
+
+	for j := uint64(0); j < nr; j++ {
+		coin := hashCoin(j, c.SigCommit, c.SignedWeight)
+		pos, err := b.coinIndex(coin, 0, uint64(len(b.sigs)))
+		if err != nil {
+			return nil, err
+		}
+
+		if pos >= uint64(len(b.participants)) {
+			return nil, fmt.Errorf("pos %d >= len(participants) %d", pos, len(b.participants))
+		}
+
+		// If we already revealed pos, no need to do it again
+		alreadyRevealed := false
+		for _, r := range c.Reveals {
+			if r.Pos == pos {
+				alreadyRevealed = true
+			}
+		}
+
+		if alreadyRevealed {
+			continue
+		}
+
```

Comment:
210-220 will require quadratic running time. No big deal at current sizes, but for some parameter settings we may end up with bigger certificates. May want to futureproof by first generating all the pos into an array of size nr, then sorting in n log n time, the deduplicating.

Reply:
I decided to store the reveals as a `map` in the `Cert` struct, instead of a list (slice).  This both addresses this point, and would have avoided the bug that Max spotted (where the previous list-based representation allowed an encoding to claim multiple reveals at the same position).